myst_nbs/case_studies/reinforcement_learning.myst.md

```diff
@@ -0,0 +1,548 @@
+---
+az.plot_posterior(data=tr, ref_val=[true_alpha, true_beta]);
+```
+
++++ {"id": "1FtAp76PBLCr"}
+
+In this example, the obtained posteriors are nicely centered around the MLE values. What we have gained is an idea of the plausible uncertainty around these values.
+
+### Alternative model using Bernoulli for the likelihood
+
+In this last section I provide an alternative implementation of the model using a Bernoulli likelihood.
+
++++
+
+:::{Note}
+One reason why it's useful to use the Bernoulli likelihood is that one can then do prior and posterior predictive sampling as well as model comparison. With `pm.Potential` you cannot do it, because PyMC does not know what is likelihood and what is prior nor how to generate random draws. Neither of this is a problem when using a `pm.Bernoulli` likelihood.
+:::
+
+```{code-cell} ipython3
+:id: pQdszDk_qYCX
+
+def right_action_probs(alpha, beta, actions, rewards):
+    rewards = at.as_tensor_variable(rewards, dtype="int32")
+    actions = at.as_tensor_variable(actions, dtype="int32")
+
+    # Compute the Qs values
+    Qs = 0.5 * at.ones((2,), dtype="float64")
+    Qs, updates = aesara.scan(
+        fn=update_Q, sequences=[actions, rewards], outputs_info=[Qs], non_sequences=[alpha]
+    )
+
+    # Apply the sotfmax transformation
+    Qs = Qs[:-1] * beta
+    logp_actions = Qs - at.logsumexp(Qs, axis=1, keepdims=True)
+
+    # Return the probabilities for the right action, in the original scale
+    return at.exp(logp_actions[:, 1])
+```
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 121
+id: S55HgqZiTfpa
+outputId: a2db2d68-8bf3-4773-8368-5b6dff310e4b
+---
+with pm.Model() as m_alt:
+    alpha = pm.Beta(name="alpha", alpha=1, beta=1)
+    beta = pm.HalfNormal(name="beta", sigma=10)
+
+    action_probs = right_action_probs(alpha, beta, actions, rewards)
+    like = pm.Bernoulli(name="like", p=action_probs, observed=actions[1:])
+
+    tr_alt = pm.sample(random_seed=rng)
+```
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 452
+id: zjXW103JiDRQ
+outputId: aafc1b1e-082e-414b-cac7-0ad805097057
+---
+az.plot_trace(data=tr_alt);
+```
+
+```{code-cell} ipython3
+:id: SDJN2w117eox
+
+az.plot_posterior(data=tr_alt, ref_val=[true_alpha, true_beta]);
+```
+
+## Watermark
+
+```{code-cell} ipython3
+%load_ext watermark
+%watermark -n -u -v -iv -w -p aeppl,xarray
+```
+
+## References
+
+:::{bibliography}
+:filter: docname in docnames
+:::
+
++++
+
+## Credits
+
+* Authored by [Ricardo Vieira](https://github.com/ricardov94) in June 2022
+
+* Adapted PyMC code from Maria Eckstein ([GitHub](https://github.com/MariaEckstein/SLCN), [PyMC Discourse](https://discourse.pymc.io/t/modeling-reinforcement-learning-of-human-participant-using-pymc3/1735))
+
+* Adapted MLE code from Robert Wilson and Anne Collins {cite:p}`collinswilson2019` ([GitHub](https://github.com/AnneCollins/TenSimpleRulesModeling))
```

Comment:
I was thinking about making these two sub-bullet points of the first one, what do you thing?

Reply:
Yes! Added in https://github.com/pymc-devs/pymc-examples/pull/410/commits/053036a309e498461f25e0f6748c281334c1f706